nats-base-client/consumermessages.ts

```diff
@@ -0,0 +1,449 @@
+          });
+        }
+      },
+    });
+
+    if (idle_heartbeat) {
+      const hbError = new NatsError(
+        Js409Errors.IdleHeartbeatMissed,
+        ErrorCode.JetStreamIdleHeartBeat,
+      );
+      this.monitor = new IdleHeartbeat(idle_heartbeat, (): boolean => {
+        // FIXME: this should be some sort of async notification of heartbeats missed
+        this._push({ isError: true, error: hbError });
+        this.resetPending()
+          .then(() => {
+          })
+          .catch(() => {
+          });
+        return false;
+      });
+    }
+
+    // now if we disconnect, the consumer could be gone
+    // or we were slow consumered by the server
+    (async () => {
+      for await (const s of c.api.nc.status()) {
+        switch (s.type) {
+          case Events.Disconnect:
+            // don't spam hb errors if we are disconnected
+            // @ts-ignore: optional chaining
+            this.monitor?.cancel();
+            break;
+          case Events.Reconnect:
+            // do some sanity checks and reset
+            // if that works resume the monitor
+            this.resetPending()
+              .then((ok) => {
+                if (ok) {
+                  // @ts-ignore: optional chaining
+                  this.monitor?.restart();
+                }
+              })
+              .catch(() => {
+                // ignored - this should have fired elsewhere
+              });
+            break;
+          default:
+            // ignored
+        }
+      }
+    })();
+
+    // this is the initial pull
+    this.pull(this.pullOptions());
+  }
+
+  _push(r: Result<JsMsg>) {
+    if (!this.callback) {
+      super.push(r);
+    } else {
+      const fn = typeof r === "function" ? r as () => void : null;
+      try {
+        if (!fn) {
+          this.callback(r);
+        } else {
+          fn();
+        }
+      } catch (err) {
+        this.callback({ isError: true, error: err });
+      }
+    }
+  }
+
+  resetPending(): Promise<boolean> {
+    // check we exist
+    return this.consumer.info()
+      .then((_ci) => {
+        // we exist, so effectively any pending state is gone
+        // so reset and re-pull
+        this.pending.msgs = 0;
+        this.pending.bytes = 0;
+        this.pending.requests = 0;
+        this.pull(this.pullOptions());
+        return true;
+      })
+      .catch((err) => {
+        // game over
+        if (err.message === "consumer not found") {
+          this.stop(err);
+        }
+        return false;
+      });
+  }
+
+  pull(opts: Partial<PullOptions>) {
+    const nc = this.consumer.api.nc;
+
+    nc.publish(
+      `${this.consumer.api.prefix}.CONSUMER.MSG.NEXT.${this.consumer.stream}.${this.consumer.name}`,
+      this.consumer.api.jc.encode(opts),
+      { reply: this.inbox },
+    );
+    this.pending.bytes += opts.max_bytes ?? 0;
+    this.pending.msgs += opts.batch ?? 0;
+    this.pending.requests++;
+  }
+
+  pullOptions(): Partial<PullOptions> {
```

Comment:
I wonder whether we should not send same request payloads for each request, e.g. for `max_messages == 100` and `threshold == 0.25` always send `batch: 25`. We will refill and not overshoot either way but (especially with bytes) it becomes more predictable.

Reply:
but that assumes even message sizes - in the case where the payloads are not the same, it likely would stall the client or am I misunderstanding.

Reply:
But your solution does not really prevent stalling if I understand correctly - it just sets `max_bytes` in pull request between the 25% - 100% of `max_bytes` option set by the user (assuming threshold is 0.75). Plus we don't have control over it, so it is very much possible that you send a request with 25% `max_bytes` option and the next message stalls either way... Only in that case we are never sure what the value in payload will be, so the most we can say for sure is "messages up to 25% of `max_bytes` will fit for sure, larger ones maybe will and maybe won't", which is not much of a guarantee.

By having constant `max_bytes` we can clearly document the behavior. Unless I'm misunderstanding the code or how `max_bytes` are handled by server.

Reply:
The issue for the stall is if the messages don't fit the constraints of max_bytes - at that point, the server will send a response saying that it couldn't do it, and then the client will send a new request, this time presumably for larger max_bytes as it has processed additional messages - that will continue until the client can make progress. If the request is constant, it is possible it will never be able to make progress.

Reply:
Lets chat about it